runtime/storage_test.go

```diff
@@ -5541,3 +5541,180 @@ func TestRuntimeAccountIterationMutation(t *testing.T) {
 		require.NoError(t, err)
 	})
 }
+
+func TestRuntimeTypeOrderInsignificance(t *testing.T) {
+
+	t.Parallel()
+
+	address := common.MustBytesToAddress([]byte{0x1})
+
+	newRuntime := func() (TestInterpreterRuntime, *TestRuntimeInterface) {
+		runtime := NewTestInterpreterRuntime()
+		accountCodes := map[common.Location][]byte{}
+
+		runtimeInterface := &TestRuntimeInterface{
+			Storage: NewTestLedger(nil, nil),
+			OnGetSigningAccounts: func() ([]Address, error) {
+				return []Address{address}, nil
+			},
+			OnResolveLocation: NewSingleIdentifierLocationResolver(t),
+			OnUpdateAccountContractCode: func(location common.AddressLocation, code []byte) error {
+				accountCodes[location] = code
+				return nil
+			},
+			OnGetAccountContractCode: func(location common.AddressLocation) (code []byte, err error) {
+				code = accountCodes[location]
+				return code, nil
+			},
+			OnEmitEvent: func(event cadence.Event) error {
+				return nil
+			},
+		}
+		return runtime, runtimeInterface
+	}
+
+	t.Run("intersection types", func(t *testing.T) {
+		t.Parallel()
+
+		runtime, runtimeInterface := newRuntime()
+
+		deployTx := DeploymentTransaction("Test", []byte(`
+            access(all)
+            contract Test {
+
+                access(all)
+                struct interface A {}
+
+
+                access(all)
+                struct interface B {}
+            }
+        `))
+
+		tx1 := []byte(`
+          import Test from 0x1
+
+          transaction {
+              prepare(account: auth(Storage) &Account) {
+
+                  let t1 = Type<&{Test.A, Test.B}>()
+                  let t2 = Type<&{Test.B, Test.A}>()
+
+                  let dict: {Type: Bool} = {}
+                  dict[t1] = true
+
+                  assert(dict[t1]!)
+                  assert(dict[t2]!)
+
+                  account.storage.save(dict, to: /storage/dict)
+              }
+          }
+        `)
+
+		tx2 := []byte(`
+          import Test from 0x1
+
+          transaction {
+              prepare(account: auth(Storage) &Account) {
+
+                  let t1 = Type<&{Test.A, Test.B}>()
```

Comment:
- Add a migration that migrates dictionary keys that are `Type` values, if the static type contains a restricted type (force re-hash and removal of old key / insert of new key) 
- Maybe we can force Go code to not use `==` through the Go code of embedding an incomparable, but zero size field 
    ```go
    type S struct {
        _ [0]func()
        /* ... */
    }
    ```